### PR TITLE
Feature/linux target

### PR DESCRIPTION
**Pull request by danielabdi-noaa**
Added docker capability for GFS.
Updated feature/linux-target to latest gfsv16b.

### CODE REVIEW

**Review by jqs-noaa — DISMISSED** [review]

**Review by jqs-noaa — COMMENTED** [review]
I thought we were moving all changes into this repo and no longer using a separate gfs-docker repo.  I don't believe this should be merged in.  Change #06242e1

**Comment by danielabdi-noaa** [comment]
@jqs-noaa Ok I will directly add the contents of the gfs-docker repository.

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa  - please create a subdirectory called cloud/ that contains the cloud work.

**Comment by christopherwharrop-noaa** [comment]
@jqs-noaa - Please don't approve PRs that you don't approve of :-)

**Comment by jqs-noaa** [comment]
@christopherwharrop-noaa  yep, wrong click :)

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa , @jqs-noaa - This PR is massive. I cannot properly evaluate it. Since this repository is not testable in any consistent way, much of the review relies on manual inspection of every change. This PR updates 134 files. There are some changes that don't make any sense to me. And some of the changes have nothing to do with our cloud work, but are just updates from EMC's copy of gfsv16b, which should be treated separately. 

Can we keep this around, for now, as a placeholder, but break this up into something that is more manageable?

I'll start by submitting a PR to bring in updates from EMC.

**Comment by jqs-noaa** [comment]
I agree with Chris,  There are too many changes to evaluate effectively.  Thanks.

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa / @jqs-noaa - I understand it may be difficult to break this up into small pieces due to the number of changes required to get a GNU build to function. However, if it is possible to break this up into smaller logical pieces that can stand alone, that would be very helpful.

The most important thing is to make sure all the changes are related to one another, meaning that if you were to remove any one of the changes, then it would no longer function correctly.

**Comment by danielabdi-noaa** [comment]
@christopherwharrop-noaa The first two commits are the most important ones
459a159 adds the necessary makefiles for a linux target
d48da08 is where I made source code changes. This is where I belive much review is required because I don't think adding additional makefiles or modulefiles for a given target machine affect existing gfs workflow. I was hoping source code changes would not be necessary or be minimal because we are focused on cloud but that is obviously not the case.

You would need both commits to compile for linux targets but I thought it is a better logical separation for a better review.

The rest of the commits are continuous updates from gfsv16b branch or due to I forgetting some files that needed to be there.

File: sorc/regrid_nemsio.fd/fv3_interface.f90
Comment (christopherwharrop-noaa) [inline sorc/regrid_nemsio.fd/fv3_interface.f90:65]: I do not understand the point of the changes in this file. Why are you just renaming gfs_grid to mgfs_grid?

Comment (danielabdi-noaa) [reply sorc/regrid_nemsio.fd/fv3_interface.f90:65]: @christopherwharrop-noaa  I didn't want to make that change but apparently it is needed to get the code compiling with GNU compilers. Intel allows such declarations
`type(gfs_grid)                                                       :: gfs_grid`
in which the type name and variable name are the same. GNU gfortran OTOH complained about it so I was forced to change the variable name. I don't recall the exact error message though. Maybe there are better solutions than this but I didn't invest much time finding out.

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa - Thank you for the explanation. When reviewing changes for a PR, all the changes for all commits are presented together. So, grouping things logically into commits is certainly a good idea, but it doesn't make the PR smaller, or the review any easier, overall if the PR contains multiple commits. If there's no logical way to break this up into smaller pieces that still function in isolation, then I guess we'll deal with it. But, it is generally better to issue PRs on small chunks whenever possible.

File: sorc/regrid_nemsio.fd/fv3_interface.f90
Comment (christopherwharrop-noaa) [reply sorc/regrid_nemsio.fd/fv3_interface.f90:65]: Aha! I didn't notice the type and variable were the same name. Makes sense.

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa - I am going to try to run some tests to see if I can make sure this works before we merge it. If you have any special tips for me to do that, please let me know.

**Comment by danielabdi-noaa** [comment]
I am not sure if the GSI clone from github is working properly. It takes really long to clone it and sometimes it fails too. In that case, I am going to have to move the checkout_gfs/gsi script outside of the container and copy it to the container later. Other than that it should be straightforward call to build the images
`
REPO=noagsl ./build_gnu_images.sh
`
from the cloud directory. At the moment, it does not even use the global-workflow code available on the host machine and will do a new clone in the container. I am working on integrating ability to build docker images from host's "build_all.sh" script.

**Comment by danielabdi-noaa** [comment]
Well, it seems GSI is not fully moved to github yet. There are subcomponents of GSI that are still cloned from gerrit. See this:

https://github.com/NOAA-EMC/GSI/blob/master/.gitmodules

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa - feature/gfsv16b-cloud needs to be remerged into feature/linux-target because the other PR got merged first. Can you please do that so that and then push again so that this PR contains only the new updates?

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa - I noticed the pointer inside of GSI back to VLab for two of its components and have asked Kate Friedman about it. As for the cloning of GSI, it does work for me (because I have VLab access), but it takes a really long time. I cloned it by hand and it takes a long time then too, probably because of the fix files that it's getting from VLab.

**Comment by christopherwharrop-noaa** [comment]
> I am working on integrating ability to build docker images from host's "build_all.sh" script.

Let's keep this PR to the bare minimum needed to get it done. Start new work in a new feature branch that will be merged later with a new PR. Otherwise, this will get too complicated.

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa - BTW, I just did a fresh checkout of linux-target and ran the `manage_externals` command with it pointing to the GitHub GSI and it worked just fine. It didn't even take very long. Maybe there was some kind of transient problem yesterday? It's working fine for me this morning.

File: cloud/build_gnu_images.sh
Comment (christopherwharrop-noaa) [inline cloud/build_gnu_images.sh:1]: Maybe more this, and the other .sh scripts in `cloud/`, into `cloud/scripts/` to keep things a bit more clean?

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa - I need to test this PR before I approve. I'm going to do a test of the normal Intel build on Hera to make sure that works as expected. And then, I'm going to do a test in a virtual Ubuntu machine on my mac laptop to see if the linux target will build there with gcc. I don't have VPN set up for accessing the Jenkins machine where we have access to Intel for the linux target, so maybe @jqs-noaa could try that to make sure it works? Once all of those tests work, and the container creation and running is independently verified, we  can approve.

**Review by jqs-noaa — COMMENTED** [review]
I'm building on jenkins-hpc to test the intel version.  A couple of items:

1. This build_intel_images.sh only builds the fv3gfs.  Will this build the whole global workflow eventually?
2. If the scripts used within docker fail for any reason, docker will continue and tag an image. Need to check script status and kill docker build process if internal build script fails.
3.  What defines success?  Scripts result in a docker container, but may be missing an executable or library.  I plan on running the C96 test case to verify build is complete.

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa / @jqs-noaa - Things are in pretty bad shape. I can't even get the official feature/gfsv16b from EMC to build anymore on Hera because of the `/contrib` module use change that was enacted. There are lots of hard-coded module settings buried deep inside a lot of nested repositories and I'm tracking them down, but I'm not there yet. Just another illustration of how badly designed this thing is. It'll take me some time, but I'll put together a patch file and issue a PR for it once I get that figured out.

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa - When I said "bad shape" I was referring to the official feature/gfsv16b. That's        not your fault. But it's a problem we have to resolve before we can even test your stuff.

**Comment by danielabdi-noaa** [comment]
@jqs-noaa @christopherwharrop-noaa The build_intel_images script should build 5 separate containers, namely, netcdf, esmf, fv3, nceplibs and gfs. I am sure the first four will build fine and you can run C96 with the fv3 image, but the current status of GSI on github still requires Vlab (gerrit access). And this prevents from fully cloning GFS (global-workflow). It will still compile the other components of global-workflow though but you have to manually save the image in that case. Or I could revert back the change I made to move global-workflow cloning to the image build stage. (i.e. clone on the host with vlab access and copy to the container).

**Comment by danielabdi-noaa** [comment]
@christopherwharrop-noaa You can still build the images on your mac if you have docker installed. The docker files will pull an Ubuntu 18.04 (for GNU) or a Centos 8 (for Intel) image before starting to build global-workflow.

**Comment by jqs-noaa** [comment]
@danielabdi-noaa The method for building intel images needs a little error protection.  If the process is interrupted, a docker image with the name 'temp' still exists will cause next attempt to build to fail.   The script should check if an image with the name 'temp' exists and remove it before continuing. 

This maybe required to build with intel, however the method of creating a 'temp' image results in docker not being able to use cache to rebuild images.  

build_netcdf.sh and build_esmf.sh could be optimized to use parallel builds (ie make -jX), currently they use single build job.

**Comment by jqs-noaa** [comment]
Building doesn't complete on jenkins-hpc using build_intel_images.sh.  

I have several errors.

1. Building NCEPLIBS:
`make[1]: Leaving directory '/opt/NCEPlibs-20190820/exec_linux.intel/src/w3nco/v2.0.6/src'
make  -C src/nemsio/v2.2.3/src # GLOBAL,SAR,UPP
make[1]: Entering directory '/opt/NCEPlibs-20190820/exec_linux.intel/src/nemsio/v2.2.3/src'
echo `pwd`
/opt/NCEPlibs-20190820/exec_linux.intel/src/nemsio/v2.2.3/src
mpiifort -qopenmp -O -g -fPIC -I. -c nemsio_openclose.f90
ifort: error #10106: Fatal error in /data/intel/compilers_and_libraries_2020.0.166/linux/bin/intel64/fortcom, terminated by kill signal
compilation aborted for nemsio_openclose.f90 (code 1)
make[1]: *** [Makefile:36: nemsio_openclose.o] Error 1
make[1]: Leaving directory '/opt/NCEPlibs-20190820/exec_linux.intel/src/nemsio/v2.2.3/src'
make: *** [Makefile:18: nemsio] Error 2

An error occurred building the NCEP libraries

2 Building FV3:

`Submodule path \'NEMS/tests/produtil/NCEPLIBS-pyprodutil\': checked out \'1c96952b092e8dc4da03b741ae1e8453fc9fe099\'
Submodule path \'WW3\': checked out \'c6b598dd6e95213cd9c832c766c2a923622dd4a5\'
Submodule path \'stochastic_physics\': checked out \'1745422af76d830757cd6035b6ea101e92b4cac1\'
+ cd /opt
+ ./patch_fv3.sh
./patch_fv3.sh: line 7: cd: NEMSfv3gfs: No such file or directory
patching file conf/configure.fv3.linux.intel
patching file modulefiles/linux.intel/fv3
./patch_fv3.sh: line 9: cd: FV3: No such file or directory
can\'t find file to patch at input line 5
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
\--------------------------
\|diff --git fv3_cap.F90 fv3_cap.F90
\|index b9162a8..30bd783 100644
\|--- fv3_cap.F90
\|+++ fv3_cap.F90
\--------------------------
File to patch:`

This maybe related to the failure in building NCEPlibs, which leaves you in the wrong working directory.  Changing directories to /opt before checking out ufs-weather-model should fix it.

This also highlights the need to check for errors in the build scripts to make docker exit if the build doesn't work.

**Comment by danielabdi-noaa** [comment]
@jqs-noaa 
It seems you have built three images netcdf, esmf, fv3 so you could run C96 with that atleast.
I also built nceplibs on jenkins like 3 weeks ago (REPO=test is mine), I will compile again to see if anything has changed.
```
REPOSITORY              TAG                 IMAGE ID            CREATED             SIZE
jstewart/fv3-intel      latest              0697de472486        10 minutes ago      1.52GB
jstewart/esmf-intel     latest              98700ad5de62        17 minutes ago      1.04GB
jstewart/netcdf-intel   latest              82c70bee81e9        47 minutes ago      860MB
test/nceplibs-intel     latest              3434c77ecf87        3 weeks ago         4.67GB
test/fv3-intel          latest              27a76fa50233        3 weeks ago         1.86GB
dshawul/fv3-intel       <none>              27a76fa50233        3 weeks ago         1.86GB
test/esmf-intel         latest              df8b14c7f03f        3 weeks ago         1.04GB
test/netcdf-intel       latest              7dfa8887ea34        3 weeks ago         860MB
ubuntu                  18.04               c3c304cb4f22        7 weeks ago         64.2MB
centos                  8                   470671670cac        4 months ago        237MB
```

**Comment by jqs-noaa** [comment]
@danielabdi-noaa 

There is an fv3 image because docker doesn't exit correctly if there is an error in the build.  The build process needs to catch these errors and exit the build process to prevent these bad containers.  If you run the image you will see there isn't any executable nor the NEMSfv3gfs directory in /opt.  The image isn't build correctly because of the errors above.

**Comment by danielabdi-noaa** [comment]
@jqs-noaa Your fv3 build does not have it for some reason. Note that nceplibs-intel is required only for GFS, and the fv3 binary clones its own copy. So even if nceplibs-intel fails, fv3 should have been built properly. I will check and let you know.

**Comment by danielabdi-noaa** [comment]
@jqs-noaa @christopherwharrop-noaa  This is very weird. I think it maybe due to Intel compiler optimization bug ?!  I know this is hard to believe but the error message lead me to the below link that suggests this but the post is from 2011 and they are talking about v12 :

https://software.intel.com/en-us/forums/intel-fortran-compiler/topic/270647

For our nemsio_openclose.f90 file, if you compile the same file with low optimization level (-o0) it compiles fine, but with higher optimization it sort of takes a while and then prints that error message.
```
[root@559f2ad863e2 src]# mpiifort -qopenmp -O -g -fPIC -I. -c nemsio_openclose.f90
ifort: error #10106: Fatal error in /data/intel/compilers_and_libraries_2020.0.166/linux/bin/intel64/fortcom, terminated by kill signal
compilation aborted for nemsio_openclose.f90 (code 1)
[root@559f2ad863e2 src]# mpiifort -qopenmp -O0 -g -fPIC -I. -c nemsio_openclose.f90
[root@559f2ad863e2 src]# mpiifort -qopenmp -O1 -g -fPIC -I. -c nemsio_openclose.f90
ifort: error #10106: Fatal error in /data/intel/compilers_and_libraries_2020.0.166/linux/bin/intel64/fortcom, terminated by kill signal
compilation aborted for nemsio_openclose.f90 (code 1)
[root@559f2ad863e2 src]# mpiifort -qopenmp -O2 -g -fPIC -I. -c nemsio_openclose.f90
ifort: error #10106: Fatal error in /data/intel/compilers_and_libraries_2020.0.166/linux/bin/intel64/fortcom, terminated by kill signal
compilation aborted for nemsio_openclose.f90 (code 1)
[root@559f2ad863e2 src]# mpiifort -qopenmp -g -fPIC -I. -c nemsio_openclose.f90
[root@559f2ad863e2 src]# mpiifort -qopenmp -O0 -g -fPIC -I. -c nemsio_openclose.f90
[root@559f2ad863e2 src]# mpiifort -qopenmp -g -fPIC -I. -c nemsio_openclose.f90
[root@559f2ad863e2 src]# mpiifort -qopenmp -O -g -fPIC -I. -c nemsio_openclose.f90
ifort: error #10106: Fatal error in /data/intel/compilers_and_libraries_2020.0.166/linux/bin/intel64/fortcom, terminated by kill signal
compilation aborted for nemsio_openclose.f90 (code 1)
```

What I am not sure about is what changed since 3 weeks ago. Maybe Michael updated the compiler since then... I am checking out specific commit hashes of NCEPLIBS so that can't be the reason.
More as I find out.

Daniel

**Comment by danielabdi-noaa** [comment]
Ok I pushed a temporary fix for this problem faa1cea by turning off openmp optimization when building nceplibs, and now I able to build fv3-intel and nceplibs-intel. The latter also has the copy of nceplibs used exclusively by fv3, and another one for global-workflow, so the same change is needed in both scripts. The makefiles don't have -qopenmp but the build scripts for the nceplibs bundle adds it if you specify -o 1. This is a temporary solution that I will revert later once I find out why the jenkins intel compiler is now complaining about it.

test-2 are the containers I built with the fix
```
REPOSITORY              TAG                 IMAGE ID            CREATED              SIZE
test-2/nceplibs-intel   latest              dc98f1490ffb        About a minute ago   4.67GB
test-2/fv3-intel        latest              1325e2f160ef        47 minutes ago       1.65GB
jstewart/esmf-intel     latest              98700ad5de62        6 hours ago          1.04GB
jstewart/netcdf-intel   latest              82c70bee81e9        7 hours ago          860MB
test/nceplibs-intel     latest              3434c77ecf87        3 weeks ago          4.67GB
test/fv3-intel          latest              27a76fa50233        3 weeks ago          1.86GB
```

**Comment by jqs-noaa** [comment]
I was finally able to get the build-intel-images.sh script to build all the way through.  

I also pushed changes to remove the 'temp' image if it exists, and added parallel build jobs for netcdf and esmf.

**Comment by danielabdi-noaa** [comment]
Ok that is good!  There is still one issue with one of the GSI files and intel compiler but I guess it has built libgsishrd.so by that point for it not matter. In any case, I will try to fix this because turning off openmp is not the best solution.
Thanks for the verification script!

File: sorc/fv3nc2nemsio.fd/makefile.sh
Comment (christopherwharrop-noaa) [inline sorc/fv3nc2nemsio.fd/makefile.sh]: Should this be "hera" instead of "theia"?

File: sorc/gfs_bufr.fd/meteorg.f
Comment (christopherwharrop-noaa) [inline sorc/gfs_bufr.fd/meteorg.f:1068]: What is the reason for this change?

File: sorc/regrid_nemsio.fd/fv3_interface.f90
Comment (christopherwharrop-noaa) [reply sorc/regrid_nemsio.fd/fv3_interface.f90:65]: I wonder if this is resolved in the develop branch of the ufs-weather-model repo. I would think it would have to be since they do have cmake that supports gcc.

Comment (christopherwharrop-noaa) [reply sorc/regrid_nemsio.fd/fv3_interface.f90:65]: Oh, nevermind. this is from a different place.

**Comment by christopherwharrop-noaa** [comment]
@jqs-noaa and @danielabdi-noaa: I did a look through changes for all 135 files in this PR. I didn't look too carefully at all the details for all the new makefiles, but I looked at the changes in source code and at the build scripts for the containers and code.

I think I'm getting close to issuing an approval for the merge, but I still have some more questions about how to design the building of the containers such that we are keeping them optimized (in size, etc) and also are ensuring we are always using the tags and locations in the top-level Externals.cfg for obtaining the code. It looks to me like it will be very difficult to reuse most of the build scripts in the top-level sorc/ directory to build the codes because of the way the build system for global workflow is designed and hard-coded to platforms. We will probably be stuck with adapting our own bits of bash for doing the actual builds.

What do you guys think?

File: sorc/gfs_bufr.fd/meteorg.f
Comment (danielabdi-noaa) [reply sorc/gfs_bufr.fd/meteorg.f:1068]: These minor tweaks I make are needed for GNU builds. If I don't do that I get

```
meteorg.f:1068:18:

      &         ps=psn(np)*1000,pd=pd3(np,1:levs))
                  1
Error: Rank mismatch in argument 'ps' at (1) (rank-1 and scalar)
meteorg.f:1106:50:
```

File: sorc/fv3nc2nemsio.fd/makefile.sh
Comment (danielabdi-noaa) [reply sorc/fv3nc2nemsio.fd/makefile.sh]: This came from when I was working with gfsv15 (port2hera) branch that did not know about hera, and has these safeguards.  This part of the code is removed in gfsv16b, so I need to remove the if conditions all in all. For the future though, we need to think about how auto-merging can lead to these kinds of errors. I will fix this shortly, thanks!

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa - Can you merge feature/gfsv16b-cloud into feature/linux-target and then test to make sure feature/linux-target still works?  Once that's done, I think I'm ready to accept this PR.

**Comment by danielabdi-noaa** [comment]
@christopherwharrop-noaa I have merged and tested the latest commits from gfsv16b. Do I have to merge gfsv16-cloud instead?

**Comment by christopherwharrop-noaa** [comment]
@danielabdi-noaa - Yes.  Because they are not the same thing.  We are wanting to merge feature/linux-target into feature/gfsv16b-cloud.  So, the very last step before the merge is to merge feature/gfsv16b-cloud into feature/linux-target to make absolutely sure that feature/linux-target is up-to-date with the latest changes in feature/gfsv16b-cloud.  Then do a final test of feature/linux-target to make sure it still works after the merge.  At that point you can guaranteed that feature/linux-target is both working, and will be mergable onto feature/gfsv16b-cloud without conflicts.

**Comment by danielabdi-noaa** [comment]
@christopherwharrop-noaa Ok I have merged from the cloud branch and also tested it.

**Review by christopherwharrop-noaa — APPROVED** [review]
I am approving this PR, with the following cautionary note. This PR is WAY too big. Future PRs this size will be rejected unless a convincing argument can be made that the work cannot be split into smaller pieces.  Going forward, please keep each new feature to a small, self-contained, capability.  Adding capabilities in small, testable, reviewable, increments, helps keep the repository free of bugs and mistakes. This is especially important in this case because we do not have any automated testing.

**Pull request merged**
